# **(Online Retail Transaction Analysis)**

## Objectives

* Load the Online Retail Transaction dataset into Python using pandas.
* Inspect the raw data and identify data quality issues.
* Clean and transform the dataset by handling missing values, duplicates, invalid records and incorrect data types.
* Creates useful features, including total transaction value.
* Prepare and save a cleaned dataset for exploratory data analysis, machine learning and Power BI visualisation. 

## Inputs

* Raw dataset: `Dataset/RawData/Online Retail.csv`.
* Transaction fields: `InvoiceNo`, `StockCode`, `Description`, `Quantity`, `InvoiceDate`, `UnitPrice`, `CustomerID` and `Country`.
* Python environment with Pandas and NumPy installed.
* The raw CSV is preserved and used as the source for the cleaning and transformation steps.

## Outputs

* Clean and transformed retail transaction dataset. 
* New calculated feature such as TotalPrice.
* Data quality checks and summry information.
* Cleaned CSV file saved for EDA, machine learning and Power BI.
* Documentation of the data cleaning and transformation process.

## Additional Comments

* This notebook forms the ETL stage of the Online Retail Transaction Analysis project. The raw dataset will be preserved separately from the cleaned dataset. 


---

# Change working directory

* We are assuming you will store the notebooks in a subfolder, therefore when running the notebook in the editor, you will need to change the working directory

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [33]:
import os
current_dir = os.getcwd()
current_dir

'c:\\Users\\Pdhar\\OneDrive\\Documents\\GitHub\\Online_Retail_Transaction_Analysis\\jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [34]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [35]:
current_dir = os.getcwd()
current_dir

'c:\\Users\\Pdhar\\OneDrive\\Documents\\GitHub\\Online_Retail_Transaction_Analysis'

# Section 1

Section 1 content

# Extract Data
The raw online Retail Transaction dataset is loaded from the Dataset/RawData folder using pandas.



In [ ]:
import os
import pandas as pd 
import numpy as np

In [37]:
# load the raw dataset
df = pd.read_csv(r"C:\Users\Pdhar\OneDrive\Documents\GitHub\Online_Retail_Transaction_Analysis\Dataset\RawData\Online Retail.csv")

# Display the first five raws
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom


### Initial Data Inspection

The raw Dataset has been successfully loaded into pandas DataFrame.
The df.head() output displays the five transactions records and confirms that the dataset has been read successfully.

## Descriptive Statistics 
  It is provide an overview of the numerical variables in the raw dataset. The describe() function shows the count, mean standard deviation, minimum, quartiles, and maximum values.

In [38]:
df.describe()

,Quantity,UnitPrice,CustomerID
count,541909.000000,541909.000000,541909.000000
mean,9.552250,4.611114,15287.518434
std,218.081158,96.759853,1484.746041
min,-80995.000000,-11062.060000,12346.000000
25%,1.000000,1.250000,14367.000000
50%,3.000000,2.080000,15287.000000
75%,10.000000,4.130000,16255.000000
max,80995.000000,38970.000000,18287.000000


### Discriptive Statistics result
'Quantity' has a mean of approximately 9.55, with values ranging from -80,995 to 80,995.
'UnitPrice' has a mean of approximately 4.61, with values ranging from -11062.06 to 38970.0 .
'CustomerID' has a mean of approximately 15,287.52 . 

## Dataset Dimentions

Identify the number of row and columns in the raw dataset.

This shows an overview of the size of dataset before any cleaning or transformation.



In [39]:
# Display the number of rows and columns

print("Number of row:",
df.shape[0])
print("Number of columns:",
      df.shape[1] )

Number of row: 541909
Number of columns: 8


## Columns names
   The columns are examined to understand the variables available in the dataset and determine which fields will be required for the analysis.

In [40]:
print("Column names:")
print(df.columns.tolist())

Column names:
['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country']


## Columns names findings
These output columns provide the necessary information for analysing retail transactions, sales performance, products, customers and patterns.

## Data types
Correct data types are important for calculation, filtering, date analysis and visualisation.  

In [41]:
df.dtypes

InvoiceNo          str
StockCode          str
Description        str
Quantity         int64
InvoiceDate        str
UnitPrice      float64
CustomerID       int64
Country            str
dtype: object

---

### Data Type Findings

'InvoiceNo', 'StockCode', 'Description', and 'Country, are comes as string values, which is appropriate and identify informatin.
 'Quantity' is stored as 'int64', which is appropriate because it represents the number of items purchased.
 'Unitprice' is store as 'float64'whic is monetary values. 
 'CustomerID' is stored as 'int64'. This will be review during cleaning process.
 'InvoiceDate, is currently store as string. It will be convert to a date time format.


## Missing Values

Missing values are check to identify incomplete records and determine which columns require cleaning before analysis.

In [42]:
df.isnull().sum()

InvoiceNo         0
StockCode         0
Description    1454
Quantity          0
InvoiceDate       0
UnitPrice         0
CustomerID        0
Country           0
dtype: int64

### Missing Values

The result shows that most of the columns contain no missing values. 
However, 'Description' column contains 1454 missing values.



## Duplicate Rows

   Duplicate rows are checked to identify repeated transaction records that could affect the accuracy of the analysis.

In [43]:
df.duplicated().sum()

np.int64(5268)

### Duplicate Rows result
    The dataset contains 5268 duplicate rows. These duplicate records represent repeated transaction and affect the accuracy of the analysis. It will be remove during data cleaning.

## Negative and Zero quantities
   The 'Quantity' column is checked for zero and negative values because they may represent returns, cancellations, or invalid transaction records.

In [44]:
print("Zero quantity:"),
(df["Quantity"]==0).sum()
print("Negative quantity:",
      (df["Quantity"] < 0).sum())


Zero quantity:
Negative quantity: 10624


### Negative and Zero Quantities result
    The dataset shows that no transactions with a quantity of zero. However, 10,624 record have negative quantities. 
    Negative quantities may represent returned or cancelled items.

## Invalid Transaction records

   Trasaction records are checked for invalid values that may affect the accuracy of the analysis. In particular, the 'UnitPrice' column is check for zero and negative values because a valid sales transaction should normally have a positive unit price. 

In [45]:
print("Zero unit price:",
(df["UnitPrice"] == 0).sum())
print("Negative unit price:",
(df["UnitPrice"] < 0).sum())

Zero unit price: 2515
Negative unit price: 2


### Invalid Unit Prices Result

    The results show that the dataset contains 2515 transactions with a zero UnitPrice and 2 transactions with a negative UnitPrice.
    A valid sales transaction should normally have a positive unit price. Therefore, these records will be treated as invalid transaction records and reviewed during the data cleaning stage.

action dates are stored in a valid date. Correct date formatting is most important for time base analysis, such as sales by day, month or year.
### Invoice Date Validation
    InvoiceDate column is check to makesure that the trans

# Section 2

## Data validation

In [46]:
import pandas as pd
df = pd.read_csv(r"C:\Users\Pdhar\OneDrive\Documents\GitHub\Online_Retail_Transaction_Analysis\Dataset\RawData\Online Retail.csv")
print("Row:", df.shape[0])
print("Columns:", df.shape[1])

Row: 541909
Columns: 8


### InvoiceDate Validation

The 'InvoiceDate'column is check to makesure that transaction dates are store in a valid date format.
Correct date is more important for time analysis.

---

In [47]:
df["InvoiceDate"] = pd.to_datetime(
    df["InvoiceDate"], 
    errors="coerce"
)
print("InvoiceDate type:",
      df["InvoiceDate"].dtype)
print("Invalid dates:",
      df["InvoiceDate"].isna().sum())

InvoiceDate type: datetime64[us]
Invalid dates: 0


### Result for InvoiceDate Validation

The InvoiceDate column was successfully change into datetime. It found 0 invalid dates are in a valid format and ready for time base analysis.

## CustomerID Validation

This column is check for missing values because a valid customer identifier is important for customer level analysis and segmentation.

In [48]:
print("Missing CustomerID:", 
df["CustomerID"].isna().sum())

Missing CustomerID: 0


### Customer ID Validation Result
    The customer id column was checked for missing values but no missing values were found in the dataset.

## Transaction Invoice Validation

   The 'InvoiceNo' column is checked for missing and cancelled transaction records. This helps ensure that each transaction has a valid invoice number before data cleaning.

In [49]:
print("Missing InvoiceNo:",)
df["InvoiceNo"].isna().sum()
print("Cancelled transactions:",
df["InvoiceNo"].astype(str).str.startswith("c").sum())

Missing InvoiceNo:
Cancelled transactions: 0


### Result of Transaction Invoice Validation
    This column was checked for missing and cancelled transactions. Missing invoice were not found. Cancelled transactions were identified and will be considered during the data cleaning process.

### Invalid Transactions Remove
    Invalid transactions removed from dataset including cancelled transactions, negative quantities and zero/negative unit prices.

In [50]:
df 
df[df["InvoiceNo"].astype(str).str.startswith("C")]

print("Rows after removing cancelled transactions:", len(df))

Rows after removing cancelled transactions: 541909


In [53]:
df = df[df["Quantity"] > 0]

print("Rows after removing invalid quantities:", len(df))

Rows after removing invalid quantities: 526054


In [55]:
df = df[df["UnitPrice"] > 0]

print("Rows after removing invalid unit prices:", len(df))

Rows after removing invalid unit prices: 524878


### Data cleaning result

* Cancelled transactions were removed
* Invalid quantities were removed
* Invalid unitprice were removed.
* Final dataset contains 526,054 rows.

---

## Data Cleaning
   The identified data quality issues are cleaned to prepare the dataset for analysis. Duplicate records, invalid quantities, invalid unit prices and cancelled transactions removed.

In [51]:
df = df.drop_duplicates()
print("Rows after removing duplicates:", len(df))

Rows after removing duplicates: 536641


### Duplicate Removal Result
    Duplicate records have been removed from the dataset. The dataset is ready for further cleaning.

## Create new feature or feature engineering

New feature create from the cleaned dataset to support sales analysis.

The TotalAmount feature is calculated by multiplying quantity by Unitprice

In [70]:
# Create TotalAmount feature
df["TotalAmount"] = df["Quantity"] * df["UnitPrice"]

print("TotalAmount feature created successfully.")
print(df[["Quantity", "UnitPrice","TotalAmount"]].head())

TotalAmount feature created successfully.
   Quantity  UnitPrice  TotalAmount
0         6       2.55        15.30
1         6       3.39        20.34
2         8       2.75        22.00
3         6       3.39        20.34
4         6       3.39        20.34


### New Feature result
The TotalAmount feature was created successfully. This feature represents the total value of each transaction and can be used for further sales analysis.

## Load Cleaned Data
Cleaned and Transformed dataset is ready to be saved for further analysis.
The dataset include new TotalAmount feature created during feature engineering.

In [73]:
# save the cleaned dataset

df.to_csv("Dataset/CleanData/Online_Retail_Cleaned.csv",
index=False) 

print("Cleaned dataset saved successfully.")

Cleaned dataset saved successfully.


# Push files to Repo

* In cases where you don't need to push files to Repo, you may replace this section with "Conclusions and Next Steps" and state your conclusions and next steps.

# Conclusions and Next Steps

The ETL process has been completed successfully.

The cleaned and transformed dataset was saved successfully and ready for exploratory data analysis and visualisation.

The next stage is to perform EDA on the cleaned dataset. This will help identify important patterns and relationships in the data.

Data Visualisations will be created to identify important sales trends, product performance, customer behaviour, and geographic patterns.

The finding from the EDA and visualisation will then be used to develop business insights and recommendations.